R/add_sda_target.R

```diff
@@ -0,0 +1,185 @@
+#'   requireNamespace("r2dii.match", quietly = TRUE)
+#' if (!installed) stop("Please install r2dii.match and r2dii.data")
+#'
+#' library(r2dii.data)
+#' library(r2dii.match)
+#'
+#' valid_matches <- match_name(loanbook_demo, ald_demo) %>%
+#'   # WARNING: Remember to validate matches (see `?prioritize`)
+#'   prioritize()
+#'
+#' valid_matches %>%
+#'   add_sda_target(
+#'     ald = ald_demo,
+#'     co2_intensity_scenario = r2dii.analysis::co2_intensity_scenario
+#'   )
+add_sda_target <- function(data, ald, co2_intensity_scenario) {
+  start_year <- co2_intensity_scenario %>%
+    select(.data$sector, .data$year) %>%
+    group_by(.data$sector) %>%
+    summarize(start_year = min(.data$year))
+
+  ald <- ald %>%
+    left_join(start_year, by = "sector") %>%
+    filter(.data$year >= .data$start_year) %>%
+    filter(!is.na(.data$emission_factor), !is.na(.data$production)) %>%
+    select(-.data$start_year)
+
+
+  ald_sda_market_benchmark <- calculate_sda_market_benchmark(
+    ald, co2_intensity_scenario
+  )
+
+  co2_scenario_with_py_and_g <- co2_intensity_scenario %>%
+    add_py_and_g_to_scenario()
+
+  target_benchmark_emission_factor <- co2_scenario_with_py_and_g %>%
+    calculate_sda_market_benchmark_target(ald_sda_market_benchmark) %>%
+    select(-.data$emission_factor_unit)
+
+  formatted_co2_intensity <- co2_scenario_with_py_and_g %>%
+    select(.data$sector, .data$year, .data$emission_factor, .data$py) %>%
+    rename(scenario_emission_factor = .data$emission_factor)
+
+  loanbook_with_weighted_emission_factors <- data %>%
+    calculate_weighted_emission_factor(ald)
+
+  loanbook_with_weighted_emission_factors %>%
+    full_join(
+      target_benchmark_emission_factor, by = c("year", "sector")
+    ) %>%
+    full_join(formatted_co2_intensity, by = c("year", "sector")) %>%
+    group_by(.data$sector) %>%
+    arrange(.data$year) %>%
+    mutate(
+      initial_portfolio_factor = first(.data$weighted_emission_factor),
+      d = .data$initial_portfolio_factor -
+        last(.data$target_weighted_emission_factor),
+      portfolio_target_emission_factor = (.data$d * .data$py) +
+        last(.data$scenario_emission_factor)
+    ) %>%
+    select(
+      .data$sector,
+      .data$year,
+      .data$weighted_emission_factor,
+      .data$portfolio_target_emission_factor,
+      .data$scenario_emission_factor
+    ) %>%
+    filter(!is.na(.data$portfolio_target_emission_factor)) %>%
+    tidyr::pivot_longer(
+      cols = tidyr::ends_with("factor"),
+      names_to = "emission_factor_name",
+      values_to = "emission_factor_value"
+    ) %>%
+    filter(!is.na(.data$emission_factor_value))
+}
+
+calculate_sda_market_benchmark <- function(market, co2_intensity_scenario) {
+  market %>%
+    group_by(.data$sector, .data$year) %>%
+    summarize(
+      sector_total_production = sum(.data$production),
+      # Alias production_weighted_emission_factor
+      x. = list(.data$production * .data$emission_factor)
```

Comment:
`list()` and `unnest()` are needless in dplyr >= 1 but best to keep them for backward compatibility.